### PR TITLE
Add ho nd wavelet

### PR DESCRIPTION
**Pull request by xueh2**
As a part of moving the nonlinear recon to mri_core, this wavelet toolbox is built.

a) Redundant dwt and idwt is supported in this toolbox. This is the most preferable wavelet in image and signal processing.

b) Fast implementation for harr dwt and idwt.

c) Tests are added to test_all for 1D to 3D wavelet transforms.

d) Current usage of dwt in gtplus is replaced with calls to this toolbox.

### CODE REVIEW

**Comment by hansenms** [comment]
Overall it looks OK. But this cannot be merged without coordination with @dchansen .

It seems we have some wavelet stuff in toolboxes/dwt and this would add a toolboxes/wavelet It has to be in one or the other location. Not both. Thoughts?

There are also some large commented sections in gtPlus that should really just be deleted. Contributes to the large number of lines added. 

@dchansen , thoughts?

**Comment by xueh2** [comment]
We could have toolbox_gpuwavelet and toolbox_cpuwavelet?

The commented code was deleted.

**Comment by hansenms** [comment]
The structure that we have is

toolboxes/dwt/gpu

and you could have added

toolboxes/dwt/cpu

So I guess my question is really was there a special reason to abandon the scheme that was already in place?

**Comment by xueh2** [comment]
Correct. I changed the target to cpudwt and move the files into toolbox/dwt/cpu.

**Comment by xueh2** [comment]
Wrong click when publishing comments.

**Comment by hansenms** [comment]
OK, lets have @dchansen give some comments/suggestions.

File: toolboxes/dwt/cpu/hoNDHarrWavelet.cpp [hunk 1]
```diff
@@ -0,0 +1,588 @@
...
+
+template<typename T>
+hoNDHarrWavelet<T>::~hoNDHarrWavelet()
+{
+}
+
+template<typename T>
+void hoNDHarrWavelet<T>::dwt1D(const T* const in, T* out, size_t RO, size_t level)
+{
+    try
```
Comment (dchansen) [inline toolboxes/dwt/cpu/hoNDHarrWavelet.cpp]: the try catch block here does nothing, as none of the operations can throw. They can segfault, but not throw.

**Comment by hansenms** [comment]
@dchansen good catch. Any other comments.

**Comment by xueh2** [comment]
Removed unnecessary try-catch blocks.

**Comment by dchansen** [comment]
A few comments:

In general, the code is nice. The basic wavelets are written very much in a C-style (memcpy and pointers everywhere), however it is quite readable, so I don't have a big issue with that. 
The headers are well documented, the testing seems nice and in short I think this is the best pull request I have seen from you @xueh2 . Nice work!

Looking forward, we should probably talk about how  merge the gtPlusOperator and the generalOperator/linearOperator used by the toolbox solvers. 

Let's merge this.

**Comment by hansenms** [comment]
@dchansen Thanks for the comments. I will merge this now. 

The next stage is exactly what you say consolidation of solvers and operators so that we don't have one set in gtplus and one for the rest. Following that consolidation there will actually be very little stuff left in gtplus and a complete merge and "harmonization" should be within reach. It is a long process but we are getting there.

**Pull request merged**